controllers/dockermachine_controller.go

```diff
@@ -158,75 +158,56 @@ func (r *DockerMachineReconciler) create(
 	}
 	r.Log.Info("Is there a cluster?", "cluster-exists", clusterExists)
 
-	controlPlanes, err := actions.ListControlPlanes(c.Name)
-	if err != nil {
-		r.Log.Error(err, "Error listing control planes")
-		return ctrl.Result{}, err
-	}
-
-	setValue := getRole(machine)
-	r.Log.Info("Creating node with role", "role", setValue)
-	if setValue == clusterAPIControlPlaneSetLabel {
-		// joining controlplane
-		if len(controlPlanes) > 0 {
+	if machine.Spec.Bootstrap.Data != nil {
+		var node *nodes.Node
+		if isControlPlaneMachine(machine) {
 			r.Log.Info("Adding a control plane node", "machine-name", machine.GetName(), "cluster-name", c.Name)
-			controlPlaneNode, err := actions.AddControlPlane(c.Name, machine.GetName(), *machine.Spec.Version)
+			node, err = actions.AddControlPlane(c.Name, machine.GetName(), *machine.Spec.Version)
 			if err != nil {
 				r.Log.Error(err, "Error adding control plane")
 				return ctrl.Result{}, err
 			}
-			providerID := actions.ProviderID(controlPlaneNode.Name())
-			dockerMachine.Spec.ProviderID = &providerID
-			dockerMachine.Status.Ready = true
-			return ctrl.Result{}, nil
-		}
-
-		r.Log.Info("Creating a new controlplane node")
-
-		if len(c.Status.APIEndpoints) == 0 {
-			r.Log.Info("Cluster has no ELB yet, waiting for cluster network to be reconciled")
-			return ctrl.Result{RequeueAfter: time.Second * 30}, nil
-		}
-
-		controlPlaneNode, err := actions.CreateControlPlane(c.Name, machine.GetName(), c.Status.APIEndpoints[0].Host, *machine.Spec.Version, nil)
-		if err != nil {
-			r.Log.Error(err, "Error creating control plane")
-			return ctrl.Result{}, err
+		} else {
+			r.Log.Info("Creating a new worker node")
+			node, err = actions.AddWorker(c.Name, machine.GetName(), *machine.Spec.Version)
+			if err != nil {
+				r.Log.Error(err, "Error creating new worker node")
+				return ctrl.Result{}, err
+			}
 		}
 		// set the machine's providerID
-		providerID := actions.ProviderID(controlPlaneNode.Name())
+		providerID := actions.ProviderID(node.Name())
 		dockerMachine.Spec.ProviderID = &providerID
 		dockerMachine.Status.Ready = true
-
-		s, err := kubeconfigToSecret(c.Name, c.Namespace)
-		if err != nil {
-			r.Log.Error(err, "Error converting kubeconfig to a secret")
-			return ctrl.Result{}, err
-		}
-		// Save the secret to the management cluster
-		if err := r.Client.Create(ctx, s); err != nil {
-			r.Log.Error(err, "Error saving secret to management cluster")
-			return ctrl.Result{}, err
-		}
 		return ctrl.Result{}, nil
 	}
+	return ctrl.Result{}, nil
+}
 
-	// If there are no control plane then we should hold off on joining workers
-	if len(controlPlanes) == 0 {
-		r.Log.Info("Sending machine back since there is no cluster to join", "machine", machine.Name)
-		return ctrl.Result{RequeueAfter: time.Second * 30}, nil
+func (r *DockerMachineReconciler) createFirstControlPlaneNode(
```

Comment:
It doesn't look like this method is used anywhere, can it be removed?

Reply:
Of course. I have deleted it.